### PR TITLE
Changes for VB-4200 - send username as a parameter to getAvailableSessions

### PR DESCRIPTION
**Pull request by dhirajshettymoj**
Changes for VB-4200 - send username as a parameter to getAvailableSessions so that reserved applications by the user for the same prisoner are ignored.

## What does this pull request do?

Fix for VB-4200
## What is the intent behind these changes?

Bug fixes.

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/ApplicationRepository.kt [hunk 1]
```diff
@@ -90,14 +90,15 @@ interface ApplicationRepository : JpaRepository<Application, Long>, JpaSpecifica
       "a.prisonerId = :prisonerId AND " +
       "a.modifyTimestamp >= :expiredDateAndTime AND " +
       "a.sessionSlotId = :sessionSlotId AND " +
-      "(:excludedApplicationReference is null OR a.reference != :excludedApplicationReference)",
-
+      "(:excludedApplicationReference is null OR a.reference != :excludedApplicationReference) AND " +
+      "(:excludeReservedApplicationsForUser is null OR a.createdBy != :excludeReservedApplicationsForUser)",
```
Comment (dhirajshettymoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/repository/ApplicationRepository.kt:94]: doing it on createdBy as a a temporary measure. Raised ticket for join here - https://dsdmoj.atlassian.net/browse/VB-4231. Will do the change to do a join as part of 4231.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/integration/IntegrationTestBase.kt [hunk 2]
```diff
@@ -393,4 +398,28 @@ abstract class IntegrationTestBase {
...
+      sessionDate = slotDate ?: sessionTemplate?.let { sessionDatesUtil.getFirstBookableSessionDay(sessionTemplate) } ?: LocalDate.now(),
+      applicationRestriction = sessionRestriction,
+      visitContact = ContactDto("John Smith", "013448811538"),
+      visitors = setOf(VisitorDto(123, true), VisitorDto(124, false)),
+      visitorSupport = ApplicationSupportDto(support),
+      actionedBy = actionedBy,
+      userType = userType,
+      allowOverBooking = allowOverBooking,
+    )
+  }
```
Comment (dhirajshettymoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/integration/IntegrationTestBase.kt:424]: moved from child class as need to use it in other classes.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/controller/VisitSessionController.kt [hunk 3]
```diff
@@ -148,8 +148,14 @@ class VisitSessionController(
...
     excludedApplicationReference: String? = null,
+    @RequestParam(value = "currentUser", required = false)
+    @Parameter(
+      description = "Username for the user making the request. Optional, ignored if not passed in.",
+      example = "user-1",
+    )
+    currentUser: String? = null,
   ): List<AvailableVisitSessionDto> {
-    return sessionService.getAvailableVisitSessions(prisonCode, prisonerId, sessionRestriction, DateRange(fromDate, toDate), excludedApplicationReference)
+    return sessionService.getAvailableVisitSessions(prisonCode, prisonerId, sessionRestriction, DateRange(fromDate, toDate), excludedApplicationReference, excludeReservedApplicationsForUser = currentUser)
```
Comment (dhirajshettymoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/controller/VisitSessionController.kt:158]: only passing parameter to getAvailableVisitSessions endpoint as do not want the stff functionality to be affected

Comment (adaviesMOJ) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/controller/VisitSessionController.kt:158]: [Suggestion]

excludeReservedApplicationsForUser seems like a boolean at glance. Could we name it something like

usernameToExcludeFromReservedApplications ?

**Review by adaviesMOJ — DISMISSED** [review]

**Review by adaviesMOJ — APPROVED** [review]

File: src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/controller/VisitSessionController.kt (hunk 3 above)
Comment (dhirajshettymoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/controller/VisitSessionController.kt]: Updated.

**Pull request merged**